### PR TITLE
add more scan dce rule tests, fix issues

### PR DESCRIPTION
**Pull request by mattjj**
* add an `instantiate` argument to `pe.dce_jaxpr`, needed for fixpoint in scan DCE rule (o/w can hit "fixpoint not reached")
* add more scan DCE testing, including checkin the jaxpr integrity and checking input/output behavior

~TODO: let's brainstorm more tests~

### CODE REVIEW

File: jax/_src/lax/control_flow.py [hunk 1]
```diff
@@ -1983,30 +1983,36 @@ def _scan_padding_rule(in_avals, out_avals, *args, jaxpr, **params):
...
+  jaxpr = eqn.params['jaxpr']
   num_consts, num_carry = eqn.params['num_consts'], eqn.params['num_carry']
+  num_xs = len(jaxpr.in_avals) - num_consts - num_carry
   used_carry_out, used_extensive_out = split_list(used_outputs, [num_carry])
   for i in range(1 + num_carry):
     used_outputs = used_carry_out + used_extensive_out
-    jaxpr, used_inputs = pe.dce_jaxpr(eqn.params['jaxpr'].jaxpr, used_outputs)
+    jaxpr_dce, used_inputs = pe.dce_jaxpr(
+        jaxpr.jaxpr, used_outputs,
+        instantiate=[False] * num_consts + used_carry_out + [False] * num_xs)
```
Comment (mattjj) [inline jax/_src/lax/control_flow.py:1994]: I had previously neglected to have an 'instantiate' here, which meant the fixedpoint might not be reached!

File: jax/interpreters/partial_eval.py [hunk 2]
```diff
@@ -1186,26 +1190,23 @@ def write(x: Atom, b: bool) -> None:
...
-      # it. Otherwise, keep the eqn unmodified and mark all inputs as used.
-      rule = dce_rules.get(eqn.primitive)
-      if rule:
-        used_ins, new_eqn = rule(used_outs, eqn)
-      else:
-        used_ins = [True] * len(eqn.invars)
-        new_eqn = eqn
-      new_eqns.append(new_eqn)
-    else:
+    if not any(used_outs) and not has_effects(eqn):
```
Comment (mattjj) [inline jax/interpreters/partial_eval.py:1200]: The changes here are just a reorganization of the logic, without changing it.

File: tests/api_test.py [hunk 3]
```diff
@@ -4502,8 +4502,31 @@ def test_pretty_print_unitvar(self):
     self.assertIn('in (*,)', str(jaxpr))
     self.assertNotIn('in (a,)', str(jaxpr))
 
-  def test_dce_jaxpr_scan(self):
-    raise unittest.SkipTest()  # TODO(mattjj)
+class DCETest(jtu.JaxTestCase):
```
Comment (mattjj) [inline tests/api_test.py:4505]: The main changes in this PR are the tests. The bug slipping through previously showed us we need better tests.

**Review by froystig — APPROVED** [review]
Nice work. Very handy comments in the added tests, too.

**Pull request merged**